cpp/src/arrow/compute/kernels/scalar_nested_test.cc

```diff
@@ -225,6 +225,423 @@ TEST(TestScalarNested, StructField) {
+        [false, 9], [true, 80]
+      ],
+      [],
+      [
+        [false, 1], [false, 2], [false, 3], [false, 4]
+      ],
+      [
+        [true, 9], [true, 2], [true, 5], [true, 8]
+      ]
+    ]
+  )";
+  auto map_array = ArrayFromJSON(map_type, input);
+  auto map_array_tweaked = TweakValidityBit(map_array, 5, false);
+
+  auto expected_all = ArrayFromJSON(list(int32()), R"(
+    [[99, null, 8], null, [67, 80], null, null, null ])");
+  auto expected_first = ArrayFromJSON(int32(), "[99, null, 67, null, null, null]");
+  auto expected_last = ArrayFromJSON(int32(), "[8, null, 80, null, null, null]");
+
+  CheckMapArrayLookupWithDifferentOptions(map_array_tweaked, true_scalar, expected_all,
+                                          expected_first, expected_last);
+}
+
+TEST_F(TestMapArrayLookupKernel, MonthDayNanoIntervalKeys) {
+  auto key_type = month_day_nano_interval();
+  auto map_type = map(key_type, utf8());
+  auto key_scalar = ScalarFromJSON(month_day_nano_interval(), R"([1, 2, -3])");
+  const char* input = R"(
+    [
+      [
+        [[-9, -10, 11], "zero"], [[1, 2, -3], "first_one"], [[11, -12, 0], "two"],
+        [[1, 2, -3], null], [[-7, -8, -9], "three"], [[1, 2, -3], "second_one"],
+        [[1, 2, -3], "last_one"]
+      ],
+      null,
+      [
+        [[-5, 6, 7], "zero_hero"], [[15, 16, 2], "almost_six"],
+        [[1, 2, -3], "the_dumb_one"], [[-7, -8, -9], "eleven"],
+        [[1, 2, -3], "the_chosen_one"], [[-5, 6, 7], "meaning of life"],
+        [[1, 2, -3], "just_one"], [[1, 2, -3], "no more ones!"]
+      ],
+      [
+        [[-5, 6, 7], "this"], [[-13, 14, -1], "has"], [[11, -12, 0], "no"],
+        [[15, 16, 2], "keys"]
+      ],
+      [
+        [[1, 2, -3], "this"], [[1, 2, -3], "should"], [[1, 2, -3], "also"],
+        [[1, 2, -3], "be"], [[1, 2, -3], "null"]
+      ],
+      []
+    ]
+  )";
+  auto map_array = ArrayFromJSON(map_type, input);
+  auto map_array_tweaked = TweakValidityBit(map_array, 4, false);
+
+  auto expected_first =
+      ArrayFromJSON(utf8(), R"(["first_one", null, "the_dumb_one", null, null, null])");
+  auto expected_last =
+      ArrayFromJSON(utf8(), R"(["last_one", null, "no more ones!", null, null, null])");
+  auto expected_all = ArrayFromJSON(list(utf8()),
+                                    R"([
+                                          ["first_one", null, "second_one", "last_one"],
+                                          null,
+                                          ["the_dumb_one", "the_chosen_one", "just_one", "no more ones!"],
+                                          null,
+                                          null,
+                                          null
+                                        ]
+                                      )");
+
+  CheckMapArrayLookupWithDifferentOptions(map_array_tweaked, key_scalar, expected_all,
+                                          expected_first, expected_last);
+}
```

Comment:
Eventually we should add a `TEST(TestMapArrayLookupKernel, Errors)` where we check error conditions like not providing a scalar or providing a scalar of the wrong type. (`ASSERT_RAISES(TypeError, MapArrayLookup(...));`)